rust/sbp/src/de.rs

```diff
@@ -174,6 +176,114 @@ impl std::fmt::Display for CrcError {
 
 impl std::error::Error for CrcError {}
 
+// pub struct Framer<R>(FramedRead<R, SbpDecoder>);
+//
+// impl<R> Framer<R> {
+//     pub fn new(reader: R) -> Self {
+//         Self(FramedRead::new(reader, Framer))
+//     }
+// }
+//
+// impl<R: io::Read> Iterator for Framer<R> {
+//     type Item = io::Result<Frame<BytesMut>>;
+//     fn next(&mut self) -> Option<Self::Item> {
+//         self.0.next()
+//     }
+// }
+
+pub struct Framer<R> {
+    inner: FramedRead<R, SbpFramer>,
+}
+
+impl<R: io::Read> Framer<R> {
+    pub fn new(reader: R) -> Self {
+        Self {
+            inner: FramedRead::new(reader, SbpFramer),
+        }
+    }
+}
+
+impl<R: io::Read> Iterator for Framer<R> {
+    type Item = Result<SbpFrame<BytesMut>, Error>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.inner.next()
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+struct SbpFramer;
+
+impl Decoder for SbpFramer {
+    type Item = SbpFrame<BytesMut>;
+    type Error = Error;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        Ok(SbpFrame::<BytesMut>::parse(src))
+    }
+    fn decode_eof(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        todo!()
+    }
+}
+
+pub struct SbpFrame<B> {
+    inner: B,
+    payload_len: usize,
+}
+
+impl SbpFrame<BytesMut> {
+    pub fn new(inner: BytesMut, payload_len: usize) -> Self {
```

Comment:
we shouldn't provide a public constructor for `Frame`. you should only be able to get them either from a `Framer` or from an existing message